### PR TITLE
Plugins should be a manifest, not an entire file

### PR DESCRIPTION
**Pull request by liztio**
Implements the changes discussed in [SEP3](https://github.com/heptio/sonobuoy/blob/master/docs/enhancements/streamlined-templates.md)

Introduces the new command `sonobuoy gen plugin e2e`

### CODE REVIEW

File: cmd/sonobuoy/app/operations/gen_plugin.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	placeholderHostname  = "<hostname>"
+	placeholderNamespace = "sonobuoy"
+)
+
+// GenPluginConfig are the input options for running
+type GenPluginConfig struct {
+	Paths      []string
+	PluginName string
+	// for testing
+	outfile io.Writer
```
Comment (chuckha) [inline cmd/sonobuoy/app/operations/gen_plugin.go]: This is a bit of a smell to me -- I think GeneratePluginManifest shouldn't worry about where it is writing bytes to. If the return type were ([]byte, error) then the writing logic could move up into the command code. 

The test for the command code could become an Example where we could capture/compare Stdout and tests for GeneratePluginManifest wouldn't have to worry about providing a Writer to the config.

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 2]
```diff
@@ -78,19 +85,46 @@ func (p *Plugin) GetResultType() string {
...
+func (p *Plugin) Run(kubeclient kubernetes.Interface, hostname string) error {
 	var (
-		b         bytes.Buffer
 		daemonSet v1beta1ext.DaemonSet
 	)
-	p.Definition.Template.Execute(&b, p.DfnTemplateData)
+	b, err := p.FillTemplate(hostname)
+	if err != nil {
+		// Already wrapped sufficiently by FillTemplate
+		return err
```
Comment (chuckha) [inline pkg/plugin/driver/daemonset/daemonset.go]: nit: it might be nice to know explicitly that FillTemplate failed vs deducing it from the final error we see wrapped up. This would also let you delete the comment

Also applies to job plugin driver

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 3]
```diff
@@ -78,19 +85,46 @@ func (p *Plugin) GetResultType() string {
 	return p.Definition.ResultType
 }
 
+//FillTemplate populates the internal Job YAML template with the values for this particular job.
+func (p *Plugin) FillTemplate(hostname string) (*bytes.Buffer, error) {
```
Comment (chuckha) [inline pkg/plugin/driver/daemonset/daemonset.go]: I'd expect the return type to be `([]byte, error)`, then line 94 & 107 could return `nil, wrapped error`.

Might be good to add a comment explaining why we use `*bytes.Buffer`.

Also applies to job driver.

File: pkg/plugin/driver/job/job.go [hunk 4]
```diff
@@ -170,8 +211,9 @@ func (p *Plugin) findPod(kubeclient kubernetes.Interface) (*v1.Pod, error) {
 	return &pods.Items[0], nil
 }
 
+// GetSessionID returns the session id associated with the plugin
 func (p *Plugin) GetSessionID() string {
-	return p.DfnTemplateData.SessionID
+	return "" //TODO(EKF) p.DfnTemplateData.SessionID
```
Comment (chuckha) [inline pkg/plugin/driver/job/job.go]: `p.SessionID`?

File: pkg/plugin/driver/job/template.go [hunk 5]
```diff
@@ -1,26 +1,25 @@
+package job
+
+import "text/template"
+
+var jobTemplate = template.Must(template.New("jobTemplate").Parse(`
```
Comment (chuckha) [inline pkg/plugin/driver/job/template.go]: this is great

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 6]
```diff
@@ -78,19 +85,46 @@ func (p *Plugin) GetResultType() string {
...
+	var b bytes.Buffer
+	// TODO (EKF): Should be YAML once we figure that out
+	container, err := utils.ContainerToJSON(&p.Definition.Spec)
```
Comment (chuckha) [inline pkg/plugin/driver/daemonset/daemonset.go]: this could replace this with `containerBytes, err := json.Marshal(&p.Definition.Spec)` and then used as bytes all the way through or turned into a string if needed.

Same applies to job driver.

File: pkg/plugin/driver/utils/utils.go [hunk 7]
```diff
@@ -89,3 +92,45 @@ func MakeErrorResult(resultType string, errdata map[string]interface{}, nodeName
...
+
+func (o *objectContainer) DeepCopyObject() kuberuntime.Object {
+	return &objectContainer{*o.DeepCopy()}
+}
+
+func (o *objectContainer) GetObjectKind() schema.ObjectKind {
+	return schema.EmptyObjectKind
+}
+
+func ContainerToYAML(container *v1.Container) (string, error) {
```
Comment (chuckha) [inline pkg/plugin/driver/utils/utils.go]: Is this used?

File: pkg/plugin/driver/utils/utils.go [hunk 8]
```diff
@@ -89,3 +92,45 @@ func MakeErrorResult(resultType string, errdata map[string]interface{}, nodeName
...
+
+	encoder := codecs.EncoderForVersion(info.Serializer, v1.SchemeGroupVersion)
+	b, err := kuberuntime.Encode(encoder, oc)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func ContainerToJSON(container *v1.Container) (string, error) {
```
Comment (chuckha) [inline pkg/plugin/driver/utils/utils.go]: this could be deleted and replaced with `json.Marshal` calls

File: pkg/plugin/loader/loader.go [hunk 9]
```diff
@@ -53,98 +51,142 @@ func LoadAllPlugins(namespace string, searchPath []string, selections []plugin.S
...
 		if err != nil {
-			return ret, err
+			return []plugin.Interface{}, errors.Wrapf(err, "couldn't scan %v for plugins", dir)
 		}
+		pluginFiles = append(pluginFiles, files...)
+	}
 
-		defns = append(defns, definitions...)
+	pluginDefs := []*pluginDefinition{}
+	for _, file := range pluginFiles {
```
Comment (chuckha) [inline pkg/plugin/loader/loader.go]: nit: the variable naming in this block is a little confusing -- maybe `pluginFiles` could be `pluginDefinitionFiles`, then `pluginDefs` could be `pluginDefinitions` and `pluginIface` could be `plugin`?

File: pkg/plugin/loader/loader.go [hunk 10]
```diff
@@ -53,98 +51,142 @@ func LoadAllPlugins(namespace string, searchPath []string, selections []plugin.S
...
-	case "DaemonSet":
-		cfg.MasterURL = "http://" + masterAddress + "/api/v1/results/by-node"
-		return daemonset.NewPlugin(namespace, dfn, cfg), nil
-	case "Job":
-		cfg.MasterURL = "http://" + masterAddress + "/api/v1/results/global"
-		return job.NewPlugin(namespace, dfn, cfg), nil
-	default:
-		return nil, errors.Errorf("Unknown driver %v", dfn.Driver)
-	}
+// func loadPlugin(namespace string, dfn plugin.Definition, masterAddress string) (plugin.Interface, error) {
```
Comment (chuckha) [inline pkg/plugin/loader/loader.go]: feel free to delete all this code or please add a comment explaining why we shouldn't/can't delete it.

File: pkg/plugin/loader/loader.go [hunk 11]
```diff
@@ -53,98 +51,142 @@ func LoadAllPlugins(namespace string, searchPath []string, selections []plugin.S
...
+// 	case "Job":
+// 		cfg.MasterURL = "http://" + masterAddress + "/api/v1/results/global"
+// 		return job.NewPlugin(namespace, dfn, cfg), nil
+// 	default:
+// 		return nil, errors.Errorf("Unknown driver %v", dfn.Driver)
+// 	}
+// }
+
+func findPlugins(dir string) ([]string, error) {
+	return filepath.Glob(path.Join(dir, "*.yml"))
```
Comment (chuckha) [inline pkg/plugin/loader/loader.go]: nit: I think kubernetes folks tend to use `*.yaml` -- but if there is a reason we need a 3 letter file extension, please comment.

File: pkg/plugin/driver/daemonset/daemonset.go (hunk 6 above)
Comment (liztio) [reply pkg/plugin/driver/daemonset/daemonset.go]: This is all superceded by a YAML output, that'll be in a subsequent PR

File: pkg/plugin/driver/utils/utils.go (hunk 7 above)
Comment (liztio) [reply pkg/plugin/driver/utils/utils.go]: nope

File: pkg/plugin/loader/loader.go (hunk 9 above)
Comment (liztio) [reply pkg/plugin/loader/loader.go]: `plugin`'s a package name, is the problem there

Comment (liztio) [reply pkg/plugin/loader/loader.go]: how's `s/pluginIface/loadedPlugin`?

File: cmd/sonobuoy/app/gen_plugin.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/heptio/sonobuoy/pkg/errlog"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+var genPluginOpts ops.GenPluginConfig
+
+func init() {
+	cmd := &cobra.Command{
+		Use:   "plugin",
```
Comment (timothysc) [inline cmd/sonobuoy/app/gen_plugin.go]: I debate if this should be a subcommand of gen and if so where should it live.  1st classing it to the root seems weird to me.

File: pkg/plugin/loader/loader.go (hunk 11 above)
Comment (liztio) [reply pkg/plugin/loader/loader.go]: let's do both!

File: cmd/sonobuoy/app/operations/gen_plugin.go (hunk 1 above)
Comment (timothysc) [reply cmd/sonobuoy/app/operations/gen_plugin.go]: I agree we should serialize to an `byte[]` which then can be processed by anything.

File: pkg/plugin/loader/loader.go (hunk 9 above)
Comment (chuckha) [reply pkg/plugin/loader/loader.go]: yeah, that sounds good to me

File: cmd/sonobuoy/app/operations/gen_plugin_test.go [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+	if err.Error() != "expected 1 plugin, got 0" {
+		t.Errorf("unexpected or no error %q", err)
+	}
+}
+
+func TestLoadRealPlugin(t *testing.T) {
+	b := &bytes.Buffer{}
+	err := GeneratePluginManifest(GenPluginConfig{
+		// Tests are executed with cwd set to their containing directory
+		Paths:      []string{"../../../../plugins.d"},
```
Comment (timothysc) [inline cmd/sonobuoy/app/operations/gen_plugin_test.go]: I'd like us to not hard code the paths if we can avoid it.  Perhaps using SONOROOT var or something.

File: pkg/plugin/loader/loader.go (hunk 11 above)
Comment (chuckha) [reply pkg/plugin/loader/loader.go]: both

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 14]
```diff
@@ -28,34 +28,41 @@ import (
...
+	Namespace  string
+	cleanedUp  bool
 }
 
 // Ensure DaemonSetPlugin implements plugin.Interface
 var _ plugin.Interface = &Plugin{}
 
+type templateData struct {
+	PluginName        string
+	ResultType        string
```
Comment (timothysc) [inline pkg/plugin/driver/daemonset/daemonset.go]: Do we even use result type anymore?  Originally it was file extension and now it's url mundge = plugin name.

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 15]
```diff
@@ -28,34 +28,41 @@ import (
...
+	SessionID         string
+	Namespace         string
+	ProducerContainer string
+	MasterAddress     string
+}
+
 // NewPlugin creates a new DaemonSet plugin from the given Plugin Definition
 // and sonobuoy master address
-func NewPlugin(namespace string, dfn plugin.Definition, cfg *plugin.WorkerConfig) *Plugin {
+func NewPlugin(dfn plugin.Definition, namespace string) *Plugin {
```
Comment (timothysc) [inline pkg/plugin/driver/daemonset/daemonset.go]: I wonder if we should continue to allow for namespace overrides.  Does that even make sense?  
We originally (knob'd everything) but I'm now re-evaluating the usefulness.

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 16]
```diff
@@ -78,19 +85,46 @@ func (p *Plugin) GetResultType() string {
...
+		// Already wrapped sufficiently by FillTemplate
+		return err
+	}
 	if err := kuberuntime.DecodeInto(scheme.Codecs.UniversalDecoder(), b.Bytes(), &daemonSet); err != nil {
 		return errors.Wrapf(err, "could not decode the executed template into a daemonset. Plugin name: ", p.GetName())
 	}
 
 	// TODO(chuckha): switch to .Apps() once extensions has been deprecated.
-	if _, err := kubeclient.ExtensionsV1beta1().DaemonSets(p.DfnTemplateData.Namespace).Create(&daemonSet); err != nil {
+	if _, err := kubeclient.ExtensionsV1beta1().DaemonSets(p.Namespace).Create(&daemonSet); err != nil {
```
Comment (timothysc) [inline pkg/plugin/driver/daemonset/daemonset.go]: eep, we should switch to apps.v1beta-something to support 1.8 & 1.9.

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 17]
```diff
@@ -110,12 +144,12 @@ func (p *Plugin) Cleanup(kubeclient kubernetes.Interface) {
 	}
 
 	// Delete the DaemonSet created by this plugin
-	err := kubeclient.ExtensionsV1beta1().DaemonSets(p.DfnTemplateData.Namespace).DeleteCollection(
+	err := kubeclient.ExtensionsV1beta1().DaemonSets(p.Namespace).DeleteCollection(
```
Comment (timothysc) [inline pkg/plugin/driver/daemonset/daemonset.go]: same.

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 18]
```diff
@@ -127,7 +161,7 @@ func (p *Plugin) listOptions() metav1.ListOptions {
 
 // findDaemonSet gets the daemonset that we created, using a kubernetes label search
 func (p *Plugin) findDaemonSet(kubeclient kubernetes.Interface) (*v1beta1ext.DaemonSet, error) {
-	dsets, err := kubeclient.ExtensionsV1beta1().DaemonSets(p.DfnTemplateData.Namespace).List(p.listOptions())
+	dsets, err := kubeclient.ExtensionsV1beta1().DaemonSets(p.Namespace).List(p.listOptions())
```
Comment (timothysc) [inline pkg/plugin/driver/daemonset/daemonset.go]: same.

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 19]
```diff
@@ -218,10 +252,14 @@ func (p *Plugin) Monitor(kubeclient kubernetes.Interface, availableNodes []v1.No
...
+	return p.SessionID
 }
 
 // GetName returns the name of this DaemonSet plugin
 func (p *Plugin) GetName() string {
 	return p.Definition.Name
 }
+
+func getMasterAddress(hostname string) string {
+	return fmt.Sprintf("http://%s/api/v1/results/global", hostname)
```
Comment (timothysc) [inline pkg/plugin/driver/daemonset/daemonset.go]: http?  not https?

File: pkg/plugin/driver/job/job.go [hunk 20]
```diff
@@ -18,45 +18,57 @@ package job
...
+	Definition plugin.Definition
+	SessionID  string
+	Namespace  string
+	cleanedUp  bool
 }
 
 // Ensure Plugin implements plugin.Interface
 var _ plugin.Interface = &Plugin{}
 
+type templateData struct {
```
Comment (timothysc) [inline pkg/plugin/driver/job/job.go]: Most of this data is repeated, couldn't we push it up a layer?

File: pkg/plugin/driver/job/job.go [hunk 21]
```diff
@@ -18,45 +18,57 @@ package job
...
-		},
-		cleanedUp: false, // be explicit
+		SessionID:  utils.GetSessionID(),
+		Namespace:  namespace,
+		cleanedUp:  false, // be explicit
 	}
 }
 
+func getMasterAddress(hostname string) string {
+	return fmt.Sprintf("http://%s/api/v1/results/by-node", hostname)
```
Comment (timothysc) [inline pkg/plugin/driver/job/job.go]: same re: https.

File: pkg/plugin/driver/job/template.go (hunk 5 above)
Comment (timothysc) [reply pkg/plugin/driver/job/template.go]: +1, I love this.

File: pkg/plugin/driver/utils/utils.go [hunk 22]
```diff
@@ -89,3 +92,45 @@ func MakeErrorResult(resultType string, errdata map[string]interface{}, nodeName
 		MimeType:   "application/json",
 	}
 }
+
```
Comment (timothysc) [inline pkg/plugin/driver/utils/utils.go]: You could probably break this into its own commit.

File: pkg/plugin/driver/daemonset/daemonset.go (hunk 19 above)
Comment (liztio) [reply pkg/plugin/driver/daemonset/daemonset.go]: not yet! This is foundational work for the HTTPS work.

File: pkg/plugin/driver/job/job.go (hunk 20 above)
Comment (liztio) [reply pkg/plugin/driver/job/job.go]: this was more of a boundary decision, I wanted every driver to have control over its own domain.

File: cmd/sonobuoy/app/gen_plugin.go (hunk 12 above)
Comment (liztio) [reply cmd/sonobuoy/app/gen_plugin.go]: I agree it's weird, but I couldn't think of any better verbs to use.

File: pkg/plugin/driver/daemonset/daemonset.go (hunk 16 above)
Comment (chuckha) [reply pkg/plugin/driver/daemonset/daemonset.go]: I think this is out of scope for this PR. I'll file a ticket.

File: pkg/plugin/driver/daemonset/daemonset.go (hunk 15 above)
Comment (liztio) [reply pkg/plugin/driver/daemonset/daemonset.go]: I think it makes sense to parameterise these even if we don't expose them to end users. That way we can have a single source of truth for the namespace

File: pkg/plugin/driver/daemonset/daemonset.go (hunk 14 above)
Comment (liztio) [reply pkg/plugin/driver/daemonset/daemonset.go]: it's used separately from the name in the worker. I don't know if anyone defines them separately, but certainly they _could_

File: cmd/sonobuoy/app/operations/gen_plugin.go [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	if len(plugins) != 1 {
+		return nil, fmt.Errorf("expected 1 plugin, got %v", len(plugins))
+	}
+
+	selectedPlugin := plugins[0]
+	bytes, err := selectedPlugin.FillTemplate(placeholderHostname)
```
Comment (chuckha) [inline cmd/sonobuoy/app/operations/gen_plugin.go]: `return selectedPlugin.FillTemplate(placeholderHostname)`

File: pkg/plugin/driver/daemonset/daemonset.go (hunk 15 above)
Comment (chuckha) [reply pkg/plugin/driver/daemonset/daemonset.go]: I believe namespaces is how heptiocloud allows users to run heptio-e2e and e2es simultaneously.

We probably want to revisit this, but that is how it was set up.

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 24]
```diff
@@ -78,19 +86,43 @@ func (p *Plugin) GetResultType() string {
 	return p.Definition.ResultType
 }
 
+//FillTemplate populates the internal Job YAML template with the values for this particular job.
+func (p *Plugin) FillTemplate(hostname string) ([]byte, error) {
+	var b bytes.Buffer
+	// TODO (EKF): Should be YAML once we figure that out
+	container, err := json.Marshal(&p.Definition.Spec)
+	if err != nil {
```
Comment (chuckha) [inline pkg/plugin/driver/daemonset/daemonset.go]: we should be consistent about `if err := fun(); err != nil` or `err := fun(); if err != nil` at least within the same function.

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 25]
```diff
@@ -78,19 +86,43 @@ func (p *Plugin) GetResultType() string {
...
-		b         bytes.Buffer
-		daemonSet v1beta1ext.DaemonSet
-	)
-	p.Definition.Template.Execute(&b, p.DfnTemplateData)
-	if err := kuberuntime.DecodeInto(scheme.Codecs.UniversalDecoder(), b.Bytes(), &daemonSet); err != nil {
+func (p *Plugin) Run(kubeclient kubernetes.Interface, hostname string) error {
+	var daemonSet appsv1beta2.DaemonSet
+
+	b, err := p.FillTemplate(hostname)
+	if err != nil {
```
Comment (chuckha) [inline pkg/plugin/driver/daemonset/daemonset.go]: move this up a line to be consistent

File: pkg/plugin/driver/job/job.go [hunk 26]
```diff
@@ -70,18 +83,47 @@ func (p *Plugin) GetResultType() string {
...
-		b   bytes.Buffer
 		job v1.Pod
 	)
-	p.Definition.Template.Execute(&b, p.DfnTemplateData)
-	if err := kuberuntime.DecodeInto(scheme.Codecs.UniversalDecoder(), b.Bytes(), &job); err != nil {
+
+	b, err := p.FillTemplate(hostname) // TODO EKF
+	if err != nil {
+		// Already wrapped sufficiently by FillTemplate
+		return err
```
Comment (chuckha) [inline pkg/plugin/driver/job/job.go]: needs a wrap & consistent style

File: pkg/plugin/interface.go [hunk 27]
```diff
@@ -41,27 +40,21 @@ type Interface interface {
 	// ExpectedResults is an array of Result objects that a plugin should
 	// expect to submit.
 	ExpectedResults(nodes []v1.Node) []ExpectedResult
+	// FillTemplate fills the driver's internal template so it can be presented to the user for `sonobuoy gen`
```
Comment (chuckha) [inline pkg/plugin/interface.go]: nit: referencing the actual application command will probably become stale in not much time

File: pkg/plugin/loader/loader.go [hunk 28]
```diff
@@ -53,98 +50,139 @@ func LoadAllPlugins(namespace string, searchPath []string, selections []plugin.S
...
 
-	return pluginDfns, err
+type pluginDefinition struct {
+	SonobuoyConfig sonobuoyConfig   `json:"sonobuoy-config"`
+	Spec           corev1.Container `json:"spec"`
 }
 
-func loadTemplate(tmpl []byte) (*plugin.Definition, error) {
-	t, err := template.New("plugin").Parse(string(tmpl))
+func loadDefinition(file string) (*pluginDefinition, error) {
```
Comment (chuckha) [inline pkg/plugin/loader/loader.go]: optional alternative: I almost want this to take an io.Reader. That way we can move the file concept out into the runner and not need to include testfiles (or mock the fs) when testing

File: pkg/plugin/driver/daemonset/daemonset.go (hunk 24 above)
Comment (liztio) [reply pkg/plugin/driver/daemonset/daemonset.go]: you can only do the latter if error is the only return, because the variables are scoped to the if statement. Unless I'm missing something?

Comment (chuckha) [reply pkg/plugin/driver/daemonset/daemonset.go]: nah u right

File: pkg/plugin/driver/daemonset/daemonset.go (hunk 25 above)
Comment (chuckha) [reply pkg/plugin/driver/daemonset/daemonset.go]: ignore

File: pkg/plugin/loader/loader.go (hunk 28 above)
Comment (liztio) [reply pkg/plugin/loader/loader.go]: We'd still need the testfiles for `findPlugins` I think, but we could certainly do this anyway

File: cmd/sonobuoy/app/gen_plugin.go (hunk 12 above)
Comment (timothysc) [reply cmd/sonobuoy/app/gen_plugin.go]: I'm ok with merging for now and revisiting the shuffle later.

File: cmd/sonobuoy/app/operations/gen_plugin.go [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+	plugins, err := loader.LoadAllPlugins(
+		placeholderNamespace,
+		cfg.Paths,
+		[]plugin.Selection{{Name: cfg.PluginName}},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(plugins) != 1 {
```
Comment (timothysc) [inline cmd/sonobuoy/app/operations/gen_plugin.go]: I'd have to grok the loadAllPlugins again, but exactly 1?

File: cmd/sonobuoy/app/operations/gen_plugin_test.go [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+
+	if err.Error() != "expected 1 plugin, got 0" {
+		t.Errorf("unexpected or no error %q", err)
+	}
+}
+
+func TestLoadRealPlugin(t *testing.T) {
+	bytes, err := GeneratePluginManifest(GenPluginConfig{
+		// Tests are executed with cwd set to their containing directory
+		Paths:      []string{"../../../../plugins.d"},
```
Comment (timothysc) [inline cmd/sonobuoy/app/operations/gen_plugin_test.go]: Could you add a //TODO to fix this later.

File: cmd/sonobuoy/app/operations/gen_plugin.go (hunk 29 above)
Comment (timothysc) [reply cmd/sonobuoy/app/operations/gen_plugin.go]: If so we should cost that value vs. magic numbering it.

File: pkg/plugin/driver/daemonset/daemonset.go [hunk 31]
```diff
@@ -78,19 +86,43 @@ func (p *Plugin) GetResultType() string {
...
+		return errors.Wrap(err, "couldn't fill template")
+	}
+	if err := kuberuntime.DecodeInto(scheme.Codecs.UniversalDecoder(), b, &daemonSet); err != nil {
 		return errors.Wrapf(err, "could not decode the executed template into a daemonset. Plugin name: ", p.GetName())
 	}
 
-	// TODO(chuckha): switch to .Apps() once extensions has been deprecated.
-	if _, err := kubeclient.ExtensionsV1beta1().DaemonSets(p.DfnTemplateData.Namespace).Create(&daemonSet); err != nil {
+	if _, err := kubeclient.AppsV1beta2().DaemonSets(p.Namespace).Create(&daemonSet); err != nil {
```
Comment (timothysc) [inline pkg/plugin/driver/daemonset/daemonset.go]: Could you add the // TODO : move to v1 in 1.10+

File: pkg/plugin/loader/loader.go [hunk 32]
```diff
@@ -53,98 +50,139 @@ func LoadAllPlugins(namespace string, searchPath []string, selections []plugin.S
...
+	decoded = convert(decoded)
+
+	jsonBytes, err := json.Marshal(decoded)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to execute template")
+		return nil, errors.Wrapf(err, "couldn't encode yaml as json for plugin definition %v", file)
+	}
+
+	var internalDef pluginDefinition
+	if err = json.Unmarshal(jsonBytes, &internalDef); err != nil {
```
Comment (timothysc) [inline pkg/plugin/loader/loader.go]: Definitely document the logic behind the decode + convert + marshal + unmarshal ... hopefully we can find our way around that later.

File: cmd/sonobuoy/app/operations/gen_plugin_test.go (hunk 30 above)
Comment (liztio) [reply cmd/sonobuoy/app/operations/gen_plugin_test.go]: I don't know how we'd go about fixing it honestly

File: pkg/plugin/loader/loader.go (hunk 32 above)
Comment (liztio) [reply pkg/plugin/loader/loader.go]: @timothysc lucky you, my next PR does away with this altogether!

File: cmd/sonobuoy/app/operations/gen_plugin.go (hunk 29 above)
Comment (liztio) [reply cmd/sonobuoy/app/operations/gen_plugin.go]: if we return more than one plugin, it's not obvious which one we should pick. Specifying exactly one removes this ambiguity.

File: cmd/sonobuoy/app/operations/gen_plugin_test.go (hunk 30 above)
Comment (timothysc) [reply cmd/sonobuoy/app/operations/gen_plugin_test.go]: Typically env-var for the root directory is passed around.

**Comment by liztio** [comment]
```
$ KUBECONFIG=./kubeconfig kubectl logs sonobuoy -n heptio-sonobuoy
time="2018-01-30T20:34:04Z" level=info msg="Scanning plugins in ./plugins.d (pwd: /)"
time="2018-01-30T20:34:04Z" level=info msg="Scanning plugins in /etc/sonobuoy/plugins.d (pwd: /)"
time="2018-01-30T20:34:04Z" level=info msg="Directory (/etc/sonobuoy/plugins.d) does not exist"
time="2018-01-30T20:34:04Z" level=info msg="Scanning plugins in ~/sonobuoy/plugins.d (pwd: /)"
time="2018-01-30T20:34:04Z" level=info msg="Directory (~/sonobuoy/plugins.d) does not exist"
time="2018-01-30T20:34:04Z" level=info msg="Filtering namespaces based on the following regex:.*|heptio-sonobuoy"
time="2018-01-30T20:34:04Z" level=info msg="Namespace default Matched=true"
time="2018-01-30T20:34:04Z" level=info msg="Namespace heptio-sonobuoy Matched=true"
time="2018-01-30T20:34:04Z" level=info msg="Namespace kube-public Matched=true"
time="2018-01-30T20:34:04Z" level=info msg="Namespace kube-system Matched=true"
time="2018-01-30T20:34:04Z" level=info msg="Namespace sonobuoy Matched=true"
time="2018-01-30T20:34:04Z" level=info msg="Starting server Expected Results: [{ e2e} {ip-10-0-18-159.ec2.internal systemd_logs} {ip-10-0-22-130.ec2.internal systemd_logs} {ip-10-0-22-8.ec2.internal systemd_logs}]"
time="2018-01-30T20:34:04Z" level=info msg="Running (e2e) plugin"
time="2018-01-30T20:34:04Z" level=info msg="starting aggregation server on 0.0.0.0:8080"
time="2018-01-30T20:34:04Z" level=info msg="Running (systemd-logs) plugin"
time="2018-01-30T20:34:25Z" level=info msg="Running non-ns query"
time="2018-01-30T20:34:25Z" level=error msg="error querying CustomResourceDefinitions: the server could not find the requested resource"
time="2018-01-30T20:34:25Z" level=info msg="Collecting Node Configuration and Health..."
time="2018-01-30T20:34:25Z" level=info msg="Creating host results for ip-10-0-18-159.ec2.internal under /tmp/sonobuoy/516c8fa9-c9a8-454d-996b-6c68ca500b61/hosts/ip-10-0-18-159.ec2.internal\n"
time="2018-01-30T20:34:25Z" level=info msg="Creating host results for ip-10-0-22-130.ec2.internal under /tmp/sonobuoy/516c8fa9-c9a8-454d-996b-6c68ca500b61/hosts/ip-10-0-22-130.ec2.internal\n"
time="2018-01-30T20:34:25Z" level=info msg="Creating host results for ip-10-0-22-8.ec2.internal under /tmp/sonobuoy/516c8fa9-c9a8-454d-996b-6c68ca500b61/hosts/ip-10-0-22-8.ec2.internal\n"
time="2018-01-30T20:34:25Z" level=info msg="Running ns query (default)"
time="2018-01-30T20:34:28Z" level=info msg="Running ns query (heptio-sonobuoy)"
time="2018-01-30T20:34:29Z" level=info msg="Collecting Pod Logs..."
time="2018-01-30T20:34:31Z" level=error msg="error querying PodPresets: the server could not find the requested resource (get podpresets.settings.k8s.io)"
time="2018-01-30T20:34:32Z" level=info msg="Running ns query (kube-public)"
time="2018-01-30T20:34:34Z" level=info msg="Running ns query (kube-system)"
time="2018-01-30T20:34:37Z" level=info msg="Running ns query (sonobuoy)"
time="2018-01-30T20:34:39Z" level=info msg="Results available at /tmp/sonobuoy/201801302034_sonobuoy_516c8fa9-c9a8-454d-996b-6c68ca500b61.tar.gz"
time="2018-01-30T20:34:39Z" level=info msg="no-exit was specified, sonobuoy is now blocking"
```

Took some tweaking but everything works now!

**Comment by timothysc** [comment]
w00t!
